usm: Fix event consumer registration

vitkyrka opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

Currently, the process monitor event stream consumer is registered
if EnableUSMEventStream is set, but there are other options which
determine if the process monitor really is used.  If the process monitor
is not used and the event stream consumer is registered, there is a
segfault when events are received since the process monitor has not been
initialized:

```
 panic: runtime error: invalid memory address or nil pointer dereference
 [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2eaba1c]
 goroutine 2168 [running]:
 pkg/network/protocols/telemetry.(*Counter).Add(...)
 pkg/process/monitor.(*EventConsumer).HandleEvent()
 pkg/security/probe.(*EventConsumer).Start.func1()
 created by pkg/security/probe.(*EventConsumer).Start in goroutine 1
```

To fix this, ensure that the conditions for registering the event stream
consumer and initializing the process monitor are the same.

### Motivation

[Segfault in staging](https://ddstaging.datadoghq.com/logs?query=service%3Asystem-probe%20status%3Aerror%20panic%20-nodegroup%3Adist-aggr_dist-aggr-c4&agg_m=count&agg_m_source=base&agg_t=count&cols=host%2Cservice&event=AgAAAZB4309hd6slhAAAAAAAAAAYAAAAAEFaQjQzMVlkQUFEV0hMVk5ScGhVYmdBZwAAACQAAAAAMDE5MDc4ZTAtYTY3OC00ZmRmLWFiNjAtNzMzMjZmZGYyNjhk&fromUser=true&messageDisplay=inline&refresh_mode=paused&storage=hot&stream_sort=desc&viz=stream&from_ts=1720014600000&to_ts=1720015200000&live=false)
https://datadoghq.atlassian.net/browse/USMON-1069

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

vitkyrka committed:
usm: Only create event consumer if used

Currently, the process monitor event stream consumer is registered
if EnableUSMEventStream is set, but there are other options which
determine if the process monitor really is used.  If the process monitor
is not used and the event stream consumer is registered, there is a
segfault when events are received since the process monitor has not been
initialized:

 panic: runtime error: invalid memory address or nil pointer dereference
 [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2eaba1c]
 goroutine 2168 [running]:
 pkg/network/protocols/telemetry.(*Counter).Add(...)
 pkg/process/monitor.(*EventConsumer).HandleEvent()
 pkg/security/probe.(*EventConsumer).Start.func1()
 created by pkg/security/probe.(*EventConsumer).Start in goroutine 1

To fix this, ensure that the conditions for registering the event stream
consumer and initializing the process monitor are the same.

vitkyrka committed:
usm: http: Build kernel_requirements for linux

Does not have any dependency on eBPF.

vitkyrka committed:
usm: Consolidate config checks

Add some helpers for config checks and consolidate them in a new package
which can be built on linux (and not just linux_bpf like the main usm
package).  These helpers will be used from the event monitor package.

DanielLavie commented on pkg/network/usm/config/config.go:
```diff
@@ -0,0 +1,25 @@
+)
+
+// NeedUSM returns true if USM is supported and enabled
+func NeedUSM(config *config.Config) bool {
```
nit: consider renaming the function to better convey its purpose of checking support and enabling status (e.g., isUSMSupportedAndEnabled)

vitkyrka replied:
Done.

vitkyrka committed:
usm: Rename NeedUSM()

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into fix-event-consumer-reg

paulcacheux commented:
The `Go Package Import Differences` shows some issue I believe

vitkyrka committed:
usm: Move Supported() helpers out of http package

The minimum requirements for USM are common to several protocols
(although a couple have separate, higher requirements), so move the
helper functions to the usmconfig package instead of the http package.

Also, this allows us to remove the dependency on the pkg/ebpf package
from the agent binary.  The agent binary got this dependencty via the
http package with the recent change to build kernel_requirements.go from
that package on linux instead of linux_bpf.

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into HEAD

vitkyrka committed:
usm: test: Fix includes after helpers move

guyarb commented on cmd/system-probe/modules/eventmonitor_linux.go:
```diff
@@ -25,6 +27,10 @@ var EventMonitor = module.Factory{
 }
 
-func createProcessMonitorConsumer(evm *eventmonitor.EventMonitor) (eventmonitor.EventConsumerInterface, error) {
+func createProcessMonitorConsumer(evm *eventmonitor.EventMonitor, config *netconfig.Config) (eventmonitor.EventConsumerInterface, error) {
+	if !usmconfig.IsUSMSupportedAndEnabled(config) || !usmconfig.NeedProcessMonitor(config) {
+		return nil, nil
+	}
+
 	return procmon.NewProcessMonitorEventConsumer(evm)
 }
```
Is it only for USM?

vitkyrka replied:
Yes, this particular consumer is only used for USM at the moment.

vitkyrka commented:
/merge

vitkyrka commented:
/merge
